File: Data_Structures/linked_list.js
// Linked List Implementation
// Linked lists - consists of nodes that are connected to each other, each containing a value and pointer to the next
// 3 things we keep track of: Head is the beginning, tail is the end, and length
// If you care about insertion and deletion (esp from beginning), it's O(1) with LL
// If you care about accessing and searching, arrays are probably better since they're O(1) but O(N) with LL
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  // Inserting at the end
  push(val) {
    let new_node = new Node(val);
    if (this.head) {
      this.tail.next = new_node;
      this.tail = new_node;
    } else {
      this.head = new_node;
      this.tail = new_node;
    }
    this.length++;
    return this;
  }
  traverse() {
    let beginning = this.head;
    while (beginning) {
      console.log(beginning.val);
      beginning = beginning.next;
    }
  }
  // Removing at the end
  // 1 -> 2 -> 3 -> 4 Point last node to null & move tail to before last
  pop() {
    if (!this.head) {
      return undefined;
    }
    let cur = this.head;
    let new_tail = this.head;
    while (cur.next) {
      new_tail = cur;
      cur = cur.next;
    }
    this.tail = new_tail;
    new_tail.next = null;
    this.length--;
    // Check if there's nothing in the list to reassign head and tail
    if (this.length == 0) {
      this.head = null;
      this.tail = null;
    }
    return cur;
  }

  // removing from the beginning
  shift() {
    if (!this.head) return undefined;
    let cur = this.head;
    let new_head = cur.next;
    this.head = new_head;

    cur.next = null;
    this.length--;
    if (this.length == 0) {
      this.tail = null;
    }
    return this;
  }

  // adding node to the beginning
  unshift(val) {
    let new_node = new Node(val);
    if (!this.head) {
      this.head = new_node;
      this.tail = new_node;
    } else {
      let cur = this.head;
      this.head = new_node;
      new_node.next = cur;
    }
    this.length++;
    return this;
  }
  // get the value of a node at an index
  get(idx) {
    let cur = this.head;
    let counter = 0;
    while (cur) {
      if (counter == idx) {
        return cur;
      }
      cur = cur.next;
      counter++;
    }
  }
  // set the value of a node at an index
  set(val, idx) {
    let found_node = this.get(idx);
    if (found_node) {
      found_node.val = val;
      return true;
    }
    return false;
  }
  // insert a node at an index
  insert(val, idx) {
    // return false if index isn't within range
    if (idx < 0 || idx >= this.length) return false;

    // insert at beginning if index is 0
    if (idx == 0) return !!this.unshift(val);

    // insert right after the idx if it's b/w 1-len(list) & insert between if other
    let new_node = new Node(val);
    let cur = this.get(idx);
    if (cur == this.tail) return !!this.push(val);

    let old_next = cur.next;
    cur.next = new_node;
    new_node.next = old_next;
    this.length++;
    return true;
  }
  // remove a node at an index
  remove(idx) {
    // return false if index isn't within range
    if (idx < 0 || idx >= this.length) return false;

    // remove at beginning if index is 0
    if (idx == 0) return !!this.shift();

    let cur = this.get(idx);
    if (cur == this.tail) return !!this.pop();
    let prev = this.get(idx - 1);
    let removed = prev.next;
    prev.next = cur.next;
    this.length--;
    return removed;
  }
  reverse() {
    // Reverse Linked List in place in O(1) space
    let node = this.head;
    this.head = this.tail;
    this.tail = node;
    let prev = null;
    let next;

    for (let i = 0; i < this.length; i++) {
      next = node.next;
      node.next = prev;

      prev = node;
      node = next;
    }

    // set last value to this.head

    // Reverse Linked List through making a list
    // let curHead = this.head;
    // let listLL = [];
    // while (curHead) {
    //   listLL.push(curHead.val);
    //   curHead = curHead.next;
    // }
    // listLL.reverse();
    // let new_LL = new Node(listLL[0]);
    // this.head = new_LL;
    // for (let i = 1; i < listLL.length; i++) {
    //   new_LL.next = new Node(listLL[i]);
    //   new_LL = new_LL.next;
    // }
    // return true;
  }
}

let list = new SinglyLinkedList(); // defining linked list

list.push("1");
list.push("2");
list.push("3");
// console.log(list.insert("999", 1));
// list.remove(1);
list.reverse();
list.traverse();

File: Data_Structures/Python/linked_list.py
'''
Linked List Implementation
Linked lists - consists of nodes that are connected to each other, each containing a value and pointer to the next
3 things we keep track of: Head is the beginning, tail is the end, and length
If you care about insertion and deletion, it's O(1) with LL. This is either O(1) or O(N) with arrays depending on if inserting in the front or back.
If you care about accessing and searching, arrays are probably better since they're O(1) but O(N) with LL
'''
# Linked List Implementation

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
    def __str__(self):
        rv = []
        node = self
        while node:
            rv.append(node.val)
            node = node.next
        return str(rv)

class SinglyLinkedList: 
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    def print(self):
        rv = []
        node = self.head
        while node:
            rv.append(node.val)
            node = node.next
        print(rv)

    # Inserting at the end
    def push(self, val):
        new_node = Node(val)
        if self.head:
            self.tail.next = new_node
            self.tail = new_node
        else:
            self.head = new_node
            self.tail = new_node
        self.length += 1
        return self

    # Remove at the end
    def pop(self):
        if self.head is None: 
            return None
        cur = self.head
        while cur.next.next:
            cur = cur.next
        self.tail = cur
        cur.next = None
        self.length -= 1
        
        if self.length == 0:
            self.head = None
            self.tail = None
        return cur

    # Remove from the beginning
    def shift(self):
        if self.head is None: 
            return None
        cur = self.head
        new_head = cur.next
        self.head = new_head
        cur.next = None
        self.length -= 1
        if (self.length == 0):
            self.tail = None
        return self
    
    # Adding from the beginning
    def unshift(self, val):
        new_node = Node(val)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            cur = self.head
            self.head = new_node
            new_node.next = cur
        self.length += 1
        return self
    
    # Get value of a node at index
    def get(self, idx):
        cur = self.head
        counter = 0
        while cur:
            if counter == idx:
                return cur
            cur = cur.next
            counter +=1
    
    # Set value of a node at index
    def set(self, val, idx):
        found_node = self.get(idx)
        if found_node:
            found_node.val = val
            return True
        return False

    # Insert node at an index
    def insert(self, val, idx):
        if idx < 0 or idx >= self.length:
            return False
        
        if idx == 0:
            return self.unshift(val) # figure out how to return boolean
        
        new_node = Node(val)
        cur = self.get(idx)
        if cur == self.tail:
            return self.push(val)
        old_next = cur.next
        cur.next = new_node
        new_node.next = old_next
        self.length += 1
        return True
    
    # Remove node at an index
    def remove(self, idx):
        if idx < 0 or idx >= self.length:
            return False
        
        if idx == 0:
            return self.shift() # figure out how to return boolean
        
        cur = self.get(idx)
        if cur == self.tail:
            return self.pop()
        prev = self.get(idx-1)
        removed = prev.next
        prev.next = cur.next
        self.length -= 1
        return removed

    # Reverse in-place in O(1) space
    def reverse(self):
        node = self.head
        self.head = self.tail
        self.tail = node
        prev = None
        
        for i in range(self.length):
            next = node.next
            node.next = prev
            prev = node
            node = next


list = SinglyLinkedList()
list.push("hello")
list.push("goodbye")
list.push("cyaa")
list.reverse()
list.print()